pylearn2/datasets/dataset.py

```diff
@@ -48,6 +48,28 @@ def iterator(self, mode=None, batch_size=None, num_batches=None,
             through the dataset and may potentially be shared by
             multiple iterator objects simultaneously (see "Notes"
             below).
+        data_specs : (space, source) pair, optional
+            `space` is an instance of `Space` (possibly a
+            `CompositeSpace`), and `source` is a string (or tuple of
+            strings if `space` is a `CompositeSpace`). `source` defines
+            which data should be returned by the iterator (for instance
+            'features', 'targets'), and `space` specifies in which
+            format it will be returned. These specifications should be
+            flat: if `space` is a `CompositeSpace`, the sub-spaces should
+            not be `CompositeSpace`s, and `source` should not be a nested
+            tuple, and the same sub-source should not be mentioned more
+            than once with the same sub-space. If `data_specs` is not
+            provided, the behaviour (which sources will be present, in
+            which order and space, or whether an Exception will be raised)
+            is not defined and may depend on the implementation of each
+            `Dataset`.
```

Comment:
`DenseDesignMatrix` supports `None`, and falls back to the `_iter_data_specs` private member for that.
As far as I recall, it was to avoid breaking existing tests.
The main benefit I can see is that the `iterator` method can be called with all default arguments, which makes it possible to use the dataset as an iterator itself, as implemented in `Dataset.__iter__()`.
I am not sure if that feature is worth having, but I think it is currently tested.